packages/SwingSet/src/lib/kmarshal.js

```diff
@@ -14,6 +14,85 @@ import { assert, Fail } from '@agoric/assert';
+  assert(desc !== undefined, 'promise lacks toStringTag getter');
+  const kref = desc.value;
+  assert.typeof(kref, 'string');
+  return kref;
+};
+
+const makeAccessorStandinPromise = kref => {
+  // TODO This Bizarro world hack is only for the version of Endo that
+  // agoric-sdk currently depends on, and is already inconsistent with
+  // the "current" endo. Once agoric-sdk depends only on the next endo,
+  // we should delete this code, and rename `makeStringStandinPromise`
+  // above to `makeStandinPromise`.
+  //
+  // Bizarro World hack for attaching a string property to a Promise, courtesy
+  // of MarkM.  Even though the @@toStringTag property nominally *is* a
+  // string, some unfortunate stuff in our hardened JS safety checks blows up
+  // if it actually is.  Eventually that will be fixed and we'll be able to
+  // use the @@toStringTag property directly, but for now we need to use this
+  // weird construct employing a sneaky getter function.  Note that this is
+  // only necessary in the first place because smallcaps encodes promise
+  // references differently from remotable object references, and the current
+  // version of the smallcaps decoder annoyingly insists that if the encoded
+  // body string says a slot is a promise, then convertSlotToVal had better by
+  // damn return an actual Promise, even if, as in the intended use case here,
+  // we neither want nor need a promise nor are capable of making any use of
+  // the fact that it is one.
+  const p = Promise.resolve(`${kref} stand in`);
+  // Bizarro World makes eslint hallucinate
+  // eslint-disable-next-line @typescript-eslint/no-floating-promises
+  Object.defineProperty(p, toStringTag, {
+    get: reveal => (reveal ? kref : NaN),
+    enumerable: false,
+  });
+  return harden(p);
+};
+
+const getAccessorStandinPromiseTag = p => {
+  // Other half of Bizarro World hack for handling promises.  Note the
+  // peculiar way the @@toStringTag getter is used.
+  const desc = Object.getOwnPropertyDescriptor(p, toStringTag);
+  assert(desc !== undefined, 'promise lacks toStringTag getter');
+
+  const getter = desc.get;
+  assert.typeof(getter, 'function', 'toStringTag getter is not a function');
+  // @ts-expect-error violates the norm that getters have zero parameters
+  const kref = getter(true);
+  assert.typeof(kref, 'string');
+  return kref;
+};
+
+const [makeStandinPromise, getStandinPromiseTag] = (() => {
+  // Use whatever works
+  try {
+    const p = makeStringStandinPromise('string mascot');
+    assert(passStyleOf(p) === 'promise');
+    return [makeStringStandinPromise, getStringStandinPromiseTag];
+  } catch (_1) {
+    try {
+      const p = makeAccessorStandinPromise('accessor mascot');
+      assert(passStyleOf(p) === 'promise');
+      return [makeAccessorStandinPromise, getAccessorStandinPromiseTag];
+    } catch (_2) {
+      throw Error('One of the promise tagging schemes should have worked');
+    }
+  }
```

Comment:
This seems like a pretty expensive way to test for which version of Endo we're on top of.  It's not really an issue for tests but possibly it could be a real perf issue in the kernel.  I'm presuming we can deem this cost acceptable mostly because we don't expect to actually pay it in production since we expect the fixed version of Endo to be in place before we actually ship this (so we've either (a) updated this code here to just use the string-valued form of @@toStringTag and gotten rid of all the try/catch hooha, or (b) we're at least running on the new Endo wherein the first try clause always succeeds so there's no notable runtime overhead).

Reply:
The expensive try/catch tests happens only at module initialization time, in order to decide whether to bind `[makeStandinPromise, getStandinPromiseTag]` to `[makeStringStandinPromise, getStringStandinPromiseTag]` or `[makeAccessorStandinPromise, getAccessorStandinPromiseTag]`. Once the module is initialized, the chosen pair of functions is then used directly with no further try/catch testing.

Reply:
Also, yes, once agoric-sdk is upgraded to an endo post https://github.com/endojs/endo/pull/1785 then the expensive tests should be removed completely and the first pair of functions should be used directly.

Reply:
Ah, I missed the setup time vs. use time distinction there.  Very good.